Adds filters for search by bank, city and state with pagination

yashy3nugu opened the pull request.

captn3m0 commented on app.rb:
```diff
@@ -22,6 +26,55 @@ def get_bank_code(code)
+      filtered_df = $df
+
+      unless state.nil?
+        filtered_df = filtered_df.where(filtered_df["STATE"].eq(state))
```
Use `ISO3166` for searching, since that is a much more sane limited set than all the misspellings that you'd see here.

yashy3nugu replied:
Our current design assumes that the frontend sends the correct state values (user selects from a list), hence the use of `ISO3166` might not be required

captn3m0 replied:
If you look at the dataset, the `ISO-3166` key is standardized and easier to search. If the frontend is anyway sending the correct value (and we're trusting it), let's ask it to send standard values instead. Otherwise, common misspellings and spelling changes in the future will break things.

yashy3nugu replied:
@captn3m0 Changed the search to use ISO3166

captn3m0 commented on app.rb:
```diff
@@ -22,6 +26,55 @@ def get_bank_code(code)
+      end
+
+      unless city.nil?
+        filtered_df = filtered_df.where(filtered_df["CITY"].eq(city))
```
No support for fuzzy search?

yashy3nugu replied:
Right now the UI design doesn't require fuzzy search, hence we have implemented only regular search (user selects from a list of options). If we were to implement fuzzy search, for which the UI could be implemented later, what do you think would be the correct approach and the effort required to do so?

captn3m0 replied:
No, experimenting with redisearch for the same. But trying to make sure that even if we switch the backend, we don't break the API.

captn3m0 replied:
There's 35K unique values for `CITY`. And the breakdown by States isn't that helpful:

```
32,ANDAMAN AND NICOBAR ISLANDS
2694,ANDHRA PRADESH
65,ARUNACHAL PRADESH
628,ASSAM
1800,BIHAR
120,CHANDIGARH
519,CHHATTISGARH
26,DADRA AND NAGAR HAVELI AND DAMAN AND DIU
161,DELHI
181,GOA
1981,GUJARAT
881,HARYANA
704,HIMACHAL PRADESH
832,JAMMU AND KASHMIR
865,JHARKHAND
2309,KARNATAKA
1746,KERALA
13,LAKSHADWEEP
1502,MADHYA PRADESH
3182,MAHARASHTRA
67,MANIPUR
133,MEGHALAYA
31,MIZORAM
46,NAGALAND
1274,ODISHA
35,PUDUCHERRY
1417,PUNJAB
2094,RAJASTHAN
54,SIKKIM
2200,TAMIL NADU
813,TELANGANA
127,TRIPURA
3130,UTTAR PRADESH
477,UTTARAKHAND
2043,WEST BENGAL
```

So for someone searching in UP, WB, KA - the dropdown for cities will be 2000+.

yashy3nugu replied:
Right now, since we have implemented pagination, we are okay with the huge amount of results. We will discuss how these results will be managed in the UI side.

yashy3nugu replied:
@captn3m0 I just had a conversation with the product manager and I was told that, the flow of the UI will be such that the user selects the filters in this order

1. Bank name
2. State
3. City
4. Branch (I will be working on this filter right now)

Even if the intermediate results are huge, the pagination should be able to divide it. Once all 4 filters are applied, the number of results would be less.

captn3m0 replied:
Switching from Bank Name to Bank Code might help a lot.

yashy3nugu replied:
Switched to bank code

captn3m0 commented on app.rb:
```diff
@@ -7,6 +7,10 @@
+require 'daru'
+
+#load the dataframe on server start
+$df = Daru::DataFrame.from_csv("IFSC.csv")
```
Any chance this can be loaded from JSON? There's a lot of effort that goes into keeping the image size minimal, and this increases doubles repo size as well.

captn3m0 replied:
Is it possible to dump the dataframe to disk? (Like how we do with the redis search). That way the JSON files aren't committed to disc, but only the compressed redis database is.

We could do something similar with keeping the CSV during build, but not for runtime

yashy3nugu replied:
@captn3m0 Dumping the dataframe to disk would result in the same size as the csv as it is the most compact version right? Maybe downloading the csv during runtime from the ifsc repo and deleting it after loading should work?

yashy3nugu replied:
@captn3m0 
Another way to obtain this dataframe would be to

1. Fetch all the keys from the redis database
2. Get the value for each key
3. Transform this data and initialize it into a dataframe

I have written a small script which can do this, but the loading into a dataframe takes about 3 mins and will need to be done everytime the server starts

captn3m0 replied:
Switching to Redis sounds good 👍🏽

captn3m0 commented on Gemfile:
```diff
@@ -8,7 +8,10 @@ gem 'redis'
 
 gem "test-unit", "~> 3.5", :group => :testing
+
+gem "matrix", "~> 0.4.2"
```
Why is this needed?

yashy3nugu replied:
This is required as daru uses the matrix library which was removed from Ruby standard library after version 3.1
https://stackoverflow.com/questions/71991514/rails-require-cannot-load-such-file-matrix

captn3m0 reviewed (changes requested):
Update the README as well, with the new search API please (and all the parameters supported).

How is the performance? What if we run a query that returns 100 results at a concurrency of 100?

Can you run a [`ab`](https://httpd.apache.org/docs/2.4/programs/ab.html) benchmark to see how it fairs with multiple concurrent queries?

yashy3nugu commented:
@captn3m0 I have updated the README and added the performance of the [`ab`](https://httpd.apache.org/docs/2.4/programs/ab.html) benchmark below. 
Performance report for `ab -n 100 -c 100 "http://localhost:9292/search?limit=100&offset=0"`
```
Server Software:        thin
Server Hostname:        localhost
Server Port:            9292

Document Path:          /search?limit=100&offset=0
Document Length:        38598 bytes

Concurrency Level:      100
Time taken for tests:   0.416 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      3918500 bytes
HTML transferred:       3859800 bytes
Requests per second:    240.24 [#/sec] (mean)
Time per request:       416.246 [ms] (mean)
Time per request:       4.162 [ms] (mean, across all concurrent requests)
Transfer rate:          9193.27 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    3   0.5      3       3
Processing:    17  263 103.5    263     396
Waiting:       14  229 110.2    204     396
Total:         17  266 103.9    266     399

Percentage of the requests served within a certain time (ms)
  50%    266
  66%    335
  75%    368
  80%    368
  90%    399
  95%    399
  98%    399
  99%    399
 100%    399 (longest request)
```

captn3m0 commented:
2 things left:

1. Switch to Bank Codes. Use [banknames.json](https://github.com/razorpay/ifsc/blob/master/src/banknames.json) to be used as the mapping.
2. Add a simple curl healthcheck to the docker container, to mitigate the 3 minute startup delay.

captn3m0 reviewed (approved):
LGTM.
